bsms.mediawiki

```diff
@@ -0,0 +1,153 @@
+* If encryption is disabled, <tt>TOKEN</tt> is set to <tt>0</tt>, and all the encryption/decryption steps below can be skipped.
+
+=====Signer=====
+
+* The Signer initiates a new secure multisig setup session by setting the <tt>TOKEN</tt>. The Signer derives an <tt>ENCRYPTION_KEY</tt> from the <tt>TOKEN</tt>. The Signer can keep the session open until a different value for the <tt>TOKEN</tt> is set.
+* The Signer generates a key record by prompting the user for a multisig derivation path and retrieves the XPUB at that derivation path. Optionally, the Signer can choose a path on behalf of the user. If the Signer chooses the path, it should try to avoid reusing XPUBs for different wallets.
+* The first line in the record must be the <tt>TOKEN</tt>. The second line must be the <tt>KEY</tt>. The <tt>KEY</tt> is an XPUB plus its key origin information, written in the descriptor-defined format, i.e.: <tt>[{master key fingerprint}/{derivation path}]{XPUB}</tt>. The third line must be a <tt>SIG</tt>, whereas <tt>SIG</tt> is the signature generated by using the private key associated with the XPUB to sign the first two lines.  The signature should follow [https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki BIP-0322], legacy format accepted. Finally, the Signer encrypts the entire record with <tt>ENCRYPTION_KEY</tt>.
+
+====Round 2====
+
+=====Coordinator=====
+
+* The Coordinator gathers key records from all participating Signers. Abort the setup if the wallet setup session has expired.
+* For each key record, the Coordinator decrypts it using <tt>ENCRYPTION_KEY</tt>. The Coordinator verifies that the included <tt>SIG</tt> is valid given the <tt>KEY</tt>.
+* If all key records look good, the Coordinator fills in all necessary information to generate a descriptor record, which is simply the descriptor string plus a <tt>CHECKSUM</tt>, all in one line. The <tt>CHECKSUM</tt> has [https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#checksums BECH32 encoding].
+* The Coordinator encrypts this descriptor record with <tt>ENCRYPTION_KEY</tt>.
+* The Coordinator sends the encrypted descriptor record to all participating Signers.
+
+=====Signer=====
+
+* The Signer imports the descriptor record, decrypts it using the <tt>ENCRYPTION_KEY</tt> derived from the open session.
+* The Signer calculates and verifies the descriptor’s <tt>CHECKSUM</tt>. Abort the setup if the <tt>CHECKSUM</tt> is incorrect.
+* The Signer checks whether one of the <tt>KEY</tt>s in the descriptor belongs to it, using path and fingerprint information included in the descriptor. The check must perform an exact match on the <tt>KEY</tt>s, and not using shortcuts such as matching fingerprints (which is trivial to spoof). Abort the setup if it doesn’t detect its own <tt>KEY</tt>.
```

Comment:
The Signer should also abort if anything in the descriptor differs from the Policy Request in `Round 1 Coordinator`

> The Coordinator constructs the multisig script and its policy parameters, such as the total number of signers and the required number of signatures

Reply:
This can be automated if the Policy Request is saved

Reply:
@Rspigler the current scheme doesn't require a "Policy Request" to be sent to the Signers in round 1.

I don't think adding that would help us here, because even if the descriptor matches the original request, the user still needs to verify the final policy in the descriptor anyway. In other words, it doesn't matter what the request was but the final result.

Reply:
I agree that the final descriptor needs to be verified regardless.

But how does the signer select what script type for xpub derivation path to generate and send over to the Coordinator, without a Policy Request? The user selects?

Reply:
In the current proposal, I specified 2 options:
(a) User could enter the derivation path themselves
(b) HWW vendor could try to make it more user-friendly by prompting for the type of wallet (singlesig vs. multisig; segwit vs. legacy vs. nested - although I think address type is made somewhat redundant after migrating to descriptors) and then the HWW would automatically choose the XPUB for you.

The user doesn't need to enter M/N params on the device. As a device owner, presumably before entering the contract they know what they are signing up for. There's also the final verification step at the end of the ceremony, where you do check for M/N.

Reply:
> presumably before entering the contract they know what they are signing up for
:+1: